CartoDB/cartodb
# User metadata export

Closes #12252 

**Limitations in this version**
- No organization-specific metadata (such as permissions). Will be tackled in #12253 
- No common data (why export + import when we can just resync it)
- No imports not syncs. Imports are not very important (mostly logging information), but we should do syncs.

**How to test**
1. Create a user
2. Create some maps, datasets, etc.
3. Write down some details (map ids, api keys, etc) to be able to compare
4. Export:
```
pg_dump -U postgres -d "cartodb_dev_user_<uuid>_db" -Z0 -Fc -f dump.sql --serializable-deferrable -v --quote-all-identifiers
Carto::UserMetadataExportService.new.export_user_to_directory(user.id, path)
```
5. Delete the user
6. Import
```
psql -U postgres -c "create role \"cartodb_staging_user_<uuid>\""
createdb -U postgres cartodb_dev_user_<uuid>_db
pg_restore -U postgres -Fc dump.sql -d cartodb_dev_user_<uuid>_db
user = Carto::UserMetadataExportService.new.import_user_from_directory(path)
```
7. Everything should be working as if nothing happened

Make sure to test that nothing broke (especially the `.carto` export/import).

Double-check that exporting and importing a .carto _doesn't_ keep ids.
